datumaro/components/hl_ops/__init__.py

```diff
@@ -0,0 +1,246 @@
+        **kwargs,
+    ) -> IDataset:
+        """
+        Applies some function to dataset items.
+
+        Results are computed lazily, if the transform supports this.
+
+        Args:
+            dataset: The dataset to be transformed
+            method: The transformation to be applied to the dataset.
+                If a string is passed, it is treated as a plugin name,
+                which is searched for in the environment
+                set by the 'env' argument
+            env: A plugin collection. If not set, the built-in plugins are used
+            **kwargs: Parameters for the transformation
+
+        Returns: a wrapper around the input dataset
+        """
+
+        if isinstance(method, str):
+            if env is None:
+                env = Environment()
+            method = env.transforms[method]
+
+        if not (inspect.isclass(method) and issubclass(method, Transform)):
+            raise TypeError(f"Unexpected 'method' argument type: {type(method)}")
+
+        produced = method(dataset, **kwargs)
+
+        return Dataset(source=produced, env=env)
+
+    @staticmethod
+    def filter(
+        dataset: IDataset,
+        expr: str,
+        *,  # pylint: disable=redefined-builtin
+        filter_annotations: bool = False,
+        remove_empty: bool = False,
+    ) -> IDataset:
+        """
+        Filters out some dataset items or annotations, using a custom filter
+        expression.
+
+        Args:
+            dataset: The dataset to be filtered
+            expr: XPath-formatted filter expression
+                (e.g. `/item[subset = 'train']`, `/item/annotation[label = 'cat']`)
+            filter_annotations: Indicates if the filter should be
+                applied to items or annotations
+            remove_empty: When filtering annotations, allows to
+                exclude empty items from the resulting dataset
+
+        Returns: a wrapper around the input dataset, which is computed lazily
+            during iteration
+        """
+
+        if filter_annotations:
+            return HLOps.transform(
+                dataset, XPathAnnotationsFilter, xpath=expr, remove_empty=remove_empty
+            )
+        else:
+            if not expr:
+                return dataset
+            return HLOps.transform(dataset, XPathDatasetFilter, xpath=expr)
+
+    @staticmethod
+    def merge(*datasets: IDataset) -> IDataset:
+        """
+        Merges several datasets using the "simple" (exact matching) algorithm:
+
+            - items are matched by (id, subset) pairs
+            - matching items share the fields available
+
+                - nothing + nothing = nothing,
+                - nothing + something = something
+                - something A + something B = conflict
+            - annotations are matched by value and shared
+            - in case of conflicts, throws an error
+
+        Returns: a wrapper around the input datasets
+        """
+
+        merger = ExactMerge()
```

Comment:
Do we only consider about `ExactMerge` for this?
Some datasets could have different categories and `ExactMerge` will be failed in this case.
So I mentioned that aggregator is going to be developed with `ExactMerge` (because subsets in a dataset will have exactly same categories), while it could be replaced with other Merger (such as UnionMerge) for the cases of merging two heterogeneous datasets. What do you think?

Reply:
The next work for ticket no. 107275 (Merge code refactoring) will deal with it. The goal of that task is to make `HLOps.merge()` tie `ExactMerge`, `UnionMerge` and `IntersectMerge` together at some higher level. For example,

```python
def merge(*dataset: IDataset, how: str = "exact"):
     if how == "exact"
           merger = ExactMerge()
     ...
```